### ISSUE TITLE
关于STAR结构中的残差链接的疑问

### ISSUE THREAD

**Issue by Aaron-yhj** [Secilia-Cxy/SOFTS#4]
作者您好，感谢您让人耳目一新的研究工作。在我阅读文章和开源代码时产生了一些疑问。
在论文的图1中，STAR结构的下方有一个‘Residual’，我理解这是残差链接。但是我对照了STAR类的代码好像并没有发现有残差链接的存在。
以下是您们所提供的关于STAR类的开源代码：

class STAR(nn.Module):
    def __init__(self, d_series, d_core):
        super(STAR, self).__init__()
        """
        STar Aggregate-Redistribute Module
        """

        self.gen1 = nn.Linear(d_series, d_series)
        self.gen2 = nn.Linear(d_series, d_core)
        self.gen3 = nn.Linear(d_series + d_core, d_series)
        self.gen4 = nn.Linear(d_series, d_series)

    def forward(self, input, *args, **kwargs):
        batch_size, channels, d_series = input.shape

        # set FFN
        combined_mean = F.gelu(self.gen1(input))
        combined_mean = self.gen2(combined_mean)

        # stochastic pooling
        if self.training:
            ratio = F.softmax(combined_mean, dim=1)
            ratio = ratio.permute(0, 2, 1)
            ratio = ratio.reshape(-1, channels)
            indices = torch.multinomial(ratio, 1)
            indices = indices.view(batch_size, -1, 1).permute(0, 2, 1)
            combined_mean = torch.gather(combined_mean, 1, indices)
            combined_mean = combined_mean.repeat(1, channels, 1)
        else:
            weight = F.softmax(combined_mean, dim=1)
            combined_mean = torch.sum(combined_mean * weight, dim=1, keepdim=True).repeat(1, channels, 1)

        # mlp fusion
        combined_mean_cat = torch.cat([input, combined_mean], -1)
        combined_mean_cat = F.gelu(self.gen3(combined_mean_cat))
        combined_mean_cat = self.gen4(combined_mean_cat)
        output = combined_mean_cat

        return output, None

combined_mean_cat = torch.cat([input, combined_mean], -1)
描述的是拼接操作。。
combined_mean_cat = F.gelu(self.gen3(combined_mean_cat))
拼接后经过线性层，并且降维，经过激活函数。
combined_mean_cat = self.gen4(combined_mean_cat)
经过激活函数后，再一次经过了线性层，维度不变。
output = combined_mean_cat
这里是我产生疑问的地方，如果按照论文中的图一所描述，这里应该是有残差链接的相关操作。不知是否是我阅读的有遗漏还是残差链接体现在别处？
请您解惑！

**Comment by Secilia-Cxy** [comment]
您好，STAR模块在EncoderLayer中被调用，您所提到的残差连接也实现在[EncoderLayer](https://github.com/Secilia-Cxy/SOFTS/blob/main/layers/Transformer_EncDec.py#L5)中。

**Comment by Aaron-yhj** [comment]
感谢您的解答，问题已解决。
